### DISCUSSION TITLE
What is a good/recommended solution for caching instance-available variables per TestClass

### DISCUSSION THREAD

**Discussion by brineryte** [pytest-dev/pytest#8369] [Q&A]
Super sorry for asking something that's probably been beaten to death, but I have a question regarding this and it's been hard to find comprehensive info about this online. Can't hurt to ask, eh? 

Scenario: I have a suite of Selenium tests using pytest. The tests also have read access to a DB to pull reliable information (instead of hard coding values) for testing scenarios. e.g. I have a test user and use queries to pull back their user id, and some item ids (which sometimes change) in preparation for certain tests.

I have my tests set up using test classes that are marked using pytest mark, for organization and a conftest with a class scoped fixture used for instantiating the webdriver for each test class. 

Inside my test classes is where my concern is. I would like to make any DB calls only once and cache the values for the duration of the TestClass being run, but still need each instance to have access to the data. I have accomplished this by utilizing another class scoped fixture per class:

```python
@pytest.fixture(scope="class")
def testClassSetup(request):
    BaseTest.__init__(request.cls, request.cls.driver)
    request.cls.log.info("Running class setup")
    request.cls.page_model= PageModel(request.cls.driver)
    # DATA STUFF
    request.cls.company= request.cls.companyData.getTestCompanyAndAccountId(TEST_COMPANY)
    request.cls.items = request.cls.itemData.getTestItemsByAccountId(request.cls.company['AccountId'], 5)

@pytest.mark.cooltest
@pytest.mark.usefixtures("classDriverSetup", "testClassSetup")
class TestClass(unittest.TestCase, BaseTest):
```
and here's my conftest fixture (for clarity):
```python
@pytest.fixture(scope="class")
def classDriverSetup(request, browser, environment):
    wdf = WebDriverFactory(browser, environment)
    driver = wdf.getWebDriverInstance()

    if request.cls is not None:
        request.cls.driver = driver

    yield driver
    driver.quit()
```

This works, but it seems maybe a little too heavy handed. Is there a better way to go about this?
pause for effect
The real reason I don't like this is...ehem....it totally wrecks intellisense in pycharm for me and it makes writing out tests more of a pain, haha. What do you think, should I just say screw it and use the typical autouse-fixture-inside-the-class-technique and just allow all the db calls? 

...(I could probably come up with another caching mechanism outside my test classes as well.... 🤔)

**Comment by portyu9** [comment]
Hi @brineryte 

For native pytest tests, I would avoid storing fixture-created state implicitly on `self`. A typed class-scoped fixture is simpler and makes the dependency explicit:

```python
from dataclasses import dataclass


@dataclass
class TestData:
    company: Company
    items: list[Item]


@pytest.fixture(scope="class")
def test_data(db) -> TestData:
    return TestData(
        company=create_company(db),
        items=create_items(db),
    )


class TestSomething:
    def test_one(self, test_data: TestData):
        assert test_data.company
```

That gives you one expensive setup per class while preserving IDE/type-checker support.

If these classes must remain `unittest.TestCase` subclasses, pytest cannot inject fixture arguments directly into the individual test methods. In that case the `request.cls` pattern you are using is reasonable.

For IDE support, declare the attributes on the base/test class:

```python
class BaseTest(unittest.TestCase):
    company: Company
    items: list[Item]
```

and let the class-scoped/autouse fixture populate those attributes.

So I would prefer explicit fixture values for native pytest tests, and class attribute injection only when compatibility with `unittest.TestCase` requires it.

Please mark this answer as accepted if it helped, thank you.
